text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,439 @@
+
+This proposal also targets documenting individual crates and their dependencies in non-cargo build systems. As will be explained, doc targets in non-cargo build systems often do not support cross-crate information.
+
+There are some files in the rustdoc output directory that are read and overwritten during every invocation of rustdoc. This proposal refers to these files as **cross-crate information**, or **CCI**, as in <https://rustc-dev-guide.rust-lang.org/rustdoc.html#multiple-runs-same-output-directory>.
+
+Build systems may run build actions in a distributed environment across separate logical filesystems. It might also be desirable to run rustdoc in a lock-free parallel mode, where every rustdoc process writes to a disjoint set of files.
+
+Cargo fully supports cross-crate information, at the cost of requiring global read-write access to the doc root (`target/doc`). There are significant scalability issues with this approach.
+
+Rustdoc needing global mutable access to the files that encode this cross-crate information has implications for caching, reproducible builds, and content hashing. By adding an option to avoid this mutation, rustdoc will serve as a first-class citizen in non-cargo build systems.
+
+These considerations motivate adding an option for outputting partial CCI (parts), which are merged (linked) with a later step.
+
+# Guide-level explanation
+
+In this example, there is a crate `t` which defines a trait `T`, and a crate `s` which defines a struct `S` that implements `T`. Our goal in this demo is for `S` to appear as an implementer in `T`'s docs, even if `s` and `t` are documented independently. This guide will be assuming that we want a crate `i` that serves as our documentation index. See the Unresolved questions section for ideas that do not require an index crate.
+
+```shell
+mkdir -p t/src s/src i/src merged/doc
+echo "pub trait T {}" > t/src/lib.rs
+echo "pub struct S; impl t::T for S {}" > s/src/lib.rs
+MERGED=file://$(realpath merged/doc)
+```
+
+[Actively use](https://doc.rust-lang.org/rustc/command-line-arguments.html#--extern-specify-where-an-external-library-is-located) `t` and `s` in `i`. The `extern crate` declarations are not needed if the crates are otherwise referenced in the index; intra-doc links are enough.
+
+```shell
+echo "extern crate t; extern crate s;" > i/src/lib.rs
+```
+
+Compile the crates.
+
+```shell
+rustc --crate-name=t --crate-type=lib --edition=2021 --emit=metadata --out-dir=t/target t/src/lib.rs
+rustc --crate-name=s --crate-type=lib --edition=2021 --emit=metadata --out-dir=s/target --extern t=t/target/libt.rmeta s/src/lib.rs
+```
+
+Document `s` and `t` independently, providing `--merge=none`, `--write-info-json`.
+
+```shell
+rustdoc \
+    -Z unstable-options \
```

Comment:
Just to be 100% clear, as this example implies, all of these flags will initially be gated behind `-Z unstable-options`, correct?

Reply:
I think so. The example scraping RFC #3123 landed flags that were behind -Zunstable-options, even though the RFC didn't mention that they would be unstable.